### ISSUE TITLE
SFTP integration: both auth and file listing

### ISSUE THREAD

**Issue by mholt** [filebrowser/filebrowser#680]
**Is your feature request related to a problem? Please describe.**
We'd like to authenticate with user accounts already configured with the SFTP server, without having to manage two sets of user accounts.

Then, depending on the user that is auth'ed, we would like to show only a certain file listing, the same listing that they would have as if they logged in via SFTP directly.

The goal here is to make SFTP available in the browser.

**Describe the solution you'd like**
FileBrowser would authenticate by trying to login to a (local/remote -- local in our case) SFTP server with the given username and password.

File browser would then show the file listing and perform operations as if they were accessing via SFTP directly.

So basically, we would like to implement an SFTP proxy of sorts. So that SFTP can be accessible via the web.

**Describe alternatives you've considered**
Not sure of any that really exist like this.

**Additional context**
Basically, we want to expose the SFTP server through a web page, and this seems like the ideal way to do it (the shortest path, as it were).

I am willing to submit a PR! Any recommendations to get started?

**Comment by mholt** [comment]
@hacdias I just updated the issue to clarify more. I realized we need more than just authentication via SFTP: we also need the file operations to be derived through SFTP; in other words, an SFTP <--> HTTPS bridge.

What do you think about that?

**Comment by hacdias** [comment]
@mholt that'll be more complicated but not _really_ complicated. We currently use Afero FS so we just need an SFTP FS compatible with that interface.

It still raises some concerns: we are saving the locations on users on the database so we'd need to have an 'SFTP' mode that solely relied on the SFTP server information for users and such.

**Comment by nambrosch** [comment]
@mholt is your backend a linux system using pam?  if so you could expose those users to apache using mod_authnz_external

**Comment by xor-gate** [comment]
@hacdias I have created the Afero SFTP FS and it should work "out of the box".

**Comment by hacdias** [comment]
@xor-gate rad! Would you be available to try adding it to File Browser? 😃

**Comment by xor-gate** [comment]
Summer is starting here, and switching jobs currently. Not sure how many time it would take me to dive into it. So I can not make any promises because it is all pro bono during my free time on open source.

**Comment by hacdias** [comment]
@xor-gate it'd be awesome if you contributed to the project :D

**Comment by unicod3** [comment]
Hello @hacdias, I would like to take a look at this if nobody works on this already.

https://github.com/spf13/afero
Currently, afero seems to have sftp fs. Is there any technical docs on how the current file system backend works? If not It would be nice if you can guide me where to look at about this.

**Comment by hacdias** [comment]
@unicod3 that would be really awesome! Right now, each user has its own `Fs` of type `afero.Fs` (https://github.com/filebrowser/filebrowser/blob/master/users/users.go#L34). So I think that with some configuration and tweaks, we could have SFTP Fs working! A per user configuration. We could even have some users with local and others with remove Fs'ses.

**Comment by unicod3** [comment]
Hello @hacdias,
I made a little demonstration for afero sftp here: https://github.com/unicod3/afero-sftp-example

Now I have some questions about the implementation, 
my first question is how should we implement it on the cli?

You mentioned about making it per user, what came to my mind as a first draft is I can implement new parameters to user add cli-command, like below:
```
filebrowser users add <username> <password> 
                                    --sftp.auth.method=<public_key_file/password> 
                                    --sftp.user=<user> 
                                    --sftp.cert=<pubkeyfile path or sftp_password>  
                                    --sftp.host=<host>  
                                    --sftp.port=<22>
```

the second question is how should we keep the credentials? In the config file, db or another way that you can suggest me to use while implementing?

**Comment by hacdias** [comment]
@unicod3 hello!

This is a good question and we should think, not only about adding SFTP, but how we want future file system backends configuration to look like. Right now, we already have something like this for authentication:

```
config set --auth.type=<auth-type> --<specific-flag>=<specific-value>
```

We could, perhaps, have something like that too:

```
config set --fs.type=<fs-type> --<specific-flag>=<specific-value>
users add <username> <password> [--fs.type=<fs-type>] --<specific-flag>=<specific-value>
```

The flags should work also on `config set` for the defaults for new users. Although this changes would, perhaps, require more work than they seem. But they sound like a good idea, no?

About the credentials, I think we should save everything on the DB as we do right now. Remember: the config file can only be used for specific server-related things, such as ports, tls certificates, etc.

**Comment by vitaliy-kuzmich** [comment]
@mholt https://github.com/browsefile/backend does 'davs' works ?

**Comment by hacdias** [comment]
I'm closing this since this repository will be archived. Please read the information on the README for more information. Thanks for all your contributions!

**Comment by kk6mrp** [comment]
The repository seems to not have been archived.
